### PR TITLE
zOS: enable new platform

### PR DESCRIPTION
**Pull request by jBarz**
Pull request to enable zOS platform.
* Platform specific compiler options
* Semaphore implementation
* uv_barrier_t definition
* Disable filesystem event tests 
* Fix unix pipe implementation.

### CODE REVIEW

File: test/test-watcher-cross-stop.c [hunk 1]
```diff
@@ -59,6 +59,10 @@ static void close_cb(uv_handle_t* handle) {
 
 
 TEST_IMPL(watcher_cross_stop) {
+#if defined(__MVS__)
+  RETURN_SKIP("IBM WORK ITEM: 130119");
```
Comment (cjihrig) [inline test/test-watcher-cross-stop.c]: What is this?

File: src/unix/thread.c [hunk 2]
```diff
@@ -302,6 +307,65 @@ int uv_sem_trywait(uv_sem_t* sem) {
   return -EINVAL;  /* Satisfy the compiler. */
 }
 
+#elif defined(__MVS__)
+
+int uv_sem_init(uv_sem_t* sem, unsigned int value) {
+  uv_sem_t a = semget(IPC_PRIVATE, 1, S_IRUSR | S_IWUSR);
```
Comment (cjihrig) [inline src/unix/thread.c]: Can you separate the variable declarations from their assignments to be consistent with the preferred project style.

File: src/unix/thread.c [hunk 3]
```diff
@@ -302,6 +307,65 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+  int b = semop(a, &s, 1); 
+  if (b == -1)
+    return -errno;
+
+  *sem = a;
+  return 0;
+}
+
+void uv_sem_destroy(uv_sem_t* sem) {
+  int a = semctl( *sem, 0, IPC_RMID);
```
Comment (cjihrig) [inline src/unix/thread.c]: Extra space after `(`.

File: src/unix/thread.c [hunk 4]
```diff
@@ -302,6 +307,65 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+
+  if (a == -1)
+    abort();
+}
+
+void uv_sem_post(uv_sem_t* sem) {
+  struct sembuf s = { 0, 1, 0 };
+  int a = semop(*sem, &s, 1);
+  if (a==-1)
```
Comment (cjihrig) [inline src/unix/thread.c]: Spaces around `==` please.

File: test/test-watcher-cross-stop.c (hunk 1 above)
Comment (jBarz) [inline test/test-watcher-cross-stop.c]: This is a reference number to a work item for the feature to be supported by the OS.

Comment (jBarz) [inline test/test-watcher-cross-stop.c]: I am going to remove this snippet and disable the test case in test/test-list.h for this platform.

File: Makefile.am [hunk 5]
```diff
@@ -281,6 +285,22 @@ if SUNOS
 test_run_tests_CFLAGS += -D__EXTENSIONS__ -D_XOPEN_SOURCE=500
 endif
 
+if OS390
+test_run_tests_CFLAGS +=  -D_UNIX03_THREADS \
```
Comment (bnoordhuis) [inline Makefile.am]: Teeny tiny nit: one space after the `+=`.

File: Makefile.am [hunk 6]
```diff
@@ -351,6 +371,30 @@ libuv_la_CFLAGS += -D__EXTENSIONS__ -D_XOPEN_SOURCE=500
 libuv_la_SOURCES += src/unix/sunos.c
 endif
 
+if OS390
+include_HEADERS += include/uv-os390.h include/pthread-fixes.h include/pthread-barrier.h
+libuv_la_CFLAGS +=  -D_UNIX03_THREADS \
```
Comment (bnoordhuis) [inline Makefile.am]: Ditto.

File: configure.ac [hunk 7]
```diff
@@ -24,10 +24,12 @@ AC_ENABLE_SHARED
 AC_ENABLE_STATIC
 AC_PROG_CC
 AM_PROG_CC_C_O
+AS_IF([AS_CASE([$host_os],[openedition*],  [false], [true])], [
+	CC_CHECK_CFLAGS_APPEND([-pedantic])
```
Comment (bnoordhuis) [inline configure.ac]: Please use two spaces.

File: include/uv-unix.h [hunk 8]
```diff
@@ -102,7 +104,9 @@ struct uv__async {
   int wfd;
 };
 
-#ifndef UV_PLATFORM_SEM_T
+#if defined(__MVS__)
+# define UV_PLATFORM_SEM_T  int
```
Comment (bnoordhuis) [inline include/uv-unix.h]: One space.

File: test/test-watcher-cross-stop.c (hunk 1 above)
Comment (bnoordhuis) [inline test/test-watcher-cross-stop.c]: The RETURN_SKIP is fine, it's that an internal work item isn't useful.  If you replace it with a more general message, it should be fine.

File: test/test-list.h [hunk 9]
```diff
@@ -165,7 +165,9 @@ TEST_DECLARE   (idle_starvation)
 TEST_DECLARE   (loop_handles)
 TEST_DECLARE   (get_loadavg)
 TEST_DECLARE   (walk_handles)
+#ifndef __MVS__
```
Comment (cjihrig) [inline test/test-list.h]: Weren't you previously skipping this in the actual test? I think that would be preferable, with a skip message that doesn't reference an internal IBM ticket.

File: src/unix/thread.c [hunk 10]
```diff
@@ -302,6 +307,78 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+  if (op_status == -1)
+    return -errno;
+
+  *sem = semid;
+  return 0;
+}
+
+void uv_sem_destroy(uv_sem_t* sem) {
+  int op_status; 
+  op_status = semctl( *sem, 0, IPC_RMID);
```
Comment (cjihrig) [inline src/unix/thread.c]: Extra space after `(`

File: src/unix/thread.c [hunk 11]
```diff
@@ -302,6 +307,78 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+
+  if (op_status == -1)
+    abort();
+}
+
+void uv_sem_post(uv_sem_t* sem) {
+  struct sembuf buf;
+  int op_status;
+
+  buf= { 0, 1, 0 };
```
Comment (cjihrig) [inline src/unix/thread.c]: Space before the `=` please.

File: src/unix/thread.c [hunk 12]
```diff
@@ -302,6 +307,78 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+
+int uv_sem_trywait(uv_sem_t* sem) {
+  struct sembuf buf;
+  int op_status;
+
+  buf = { 0, -1, IPC_NOWAIT };
+
+  do
+    op_status = semop(*sem, &s, 1);
+  while (r == -1 && errno == EINTR);
```
Comment (cjihrig) [inline src/unix/thread.c]: Should `r` be `op_status`?

Comment (jBarz) [inline src/unix/thread.c]: My bad. Should test before upstreaming.

File: Makefile.am [hunk 13]
```diff
@@ -128,7 +128,11 @@ EXTRA_DIST = test/fixtures/empty_file \
 
 TESTS = test/run-tests
 check_PROGRAMS = test/run-tests
+if OS390
+test_run_tests_CFLAGS =
```
Comment (saghul) [inline Makefile.am]: why skip the -Wno-long-long?

File: test/test-poll.c [hunk 14]
```diff
@@ -388,17 +388,17 @@ static void connection_poll_cb(uv_poll_t* handle, int status, int events) {
...
+#if !defined(_AIX) && !defined(__MVS__)
   if (events & UV_DISCONNECT) {
     context->got_disconnect = 1;
     ++disconnects;
     new_events &= ~UV_DISCONNECT;
   }
 
   if (context->got_fin && context->sent_fin && context->got_disconnect) {
-#else /* _AIX */
+#else /* _AIX  && __MVS */
```
Comment (saghul) [inline test/test-poll.c]: typo? this is the only instance on this patch where `__MVS` is used instead of `__MVS__` it seems.

File: Makefile.am [hunk 15]
```diff
@@ -351,6 +372,31 @@ libuv_la_CFLAGS += -D__EXTENSIONS__ -D_XOPEN_SOURCE=500
...
+                   -D_OPEN_SYS_SOCK_IPV6 \
+                   -D_OPEN_SYS_FILE_EXT \
+                   -DUV_PLATFORM_SEM_T=int \
+                   -DUTMP_FILE=\"/etc/utmpx\" \
+                   -DPATH_MAX=255 \
+                   -qCHARS=signed \
+                   -qXPLINK \
+                   -qFLOAT=IEEE
+libuv_la_LDFLAGS += -qXPLINK
+libuv_la_SOURCES += src/unix/os390.c \
```
Comment (saghul) [inline Makefile.am]: where are these files?

**Comment by saghul** [comment]
I can't really comment on the zOS-specific code, but left some comments.

@jBarz will IBM have a buildbot for zOS? You'll basically become the de-facto maintainer of the port, at any rate :-)

Question: any chance GYP can also be used? Right now all Unixes can be built with either GYP or autotools, not sure how much we care though.

File: Makefile.am (hunk 13 above)
Comment (jBarz) [inline Makefile.am]: the C compiler on zOS by default does not issue warnings for long-long type being used.

File: Makefile.am (hunk 15 above)
Comment (jBarz) [inline Makefile.am]: These files are not yet committed because they contain some half baked zOS implementations that will be involved in a future pull request.

File: test/test-poll.c (hunk 14 above)
Comment (jBarz) [inline test/test-poll.c]: Changed to __MVS__

**Comment by jBarz** [comment]
We are looking into securing some external zOS systems to contribute to the libuv/node community, but don’t have such systems available at the moment.   We currently are cloning and building it on zOS systems available internally at IBM.

File: Makefile.am (hunk 13 above)
Comment (saghul) [inline Makefile.am]: but does it hurt to have it? it would simplify the changes

File: Makefile.am (hunk 15 above)
Comment (saghul) [inline Makefile.am]: I'm really confused here. This won't even compile then, AFAICT. We can't merge this PR as is, with some zOS fixes which don't even compile. We can wait until it's complete. Failing tests are ok, but it should at the very least compile.

File: Makefile.am (hunk 13 above)
Comment (jBarz) [inline Makefile.am]: Forgot to also mention that the zos compiler does not recognize -W option format.
So the compile fails with "option not recognized"

File: Makefile.am (hunk 15 above)
Comment (jBarz) [inline Makefile.am]: I will remove these files from the Makefile for now.

File: Makefile.am (hunk 13 above)
Comment (saghul) [inline Makefile.am]: Got it, leave it then 👍

**Comment by jBarz** [comment]
I have squashed all extra commits into meaningful ones.

**Comment by saghul** [comment]
This patch is becoming quite invasive. One thing I'd like to know is what kind of commitment we can expect so we don't have to gut out this code in X months because it went unmaintained.

**Comment by jBarz** [comment]
cc: @mhdawson 
As we have done for Linux on PPC, AIX and Linux on s390 we are working toward Node.js support for z/OS.  As we have done for those platforms our intention is to maintain currency and to ensure that support for Node.js on z/OS. Once this is achieved and we can ship Node.js for z/OS we plan to support our customer in an ongoing manner which includes keeping Node.js and its dependencies current.

**Comment by jBarz** [comment]
I will stop pulling in more changes into this PR.

**Comment by saghul** [comment]
@jBarz That's good to hear! Please continue pushing changes here until zOS support is complete. I was just wondering if these would be future-proof, since the changeset is becoming quite big. If that is the case, I'm happy to have them.

**Comment by mhdawson** [comment]
@saghul  thanks

**Comment by jBarz** [comment]
@saghul : If it is ok with you, I will stop pulling in any more changes into this PR since it is becoming too big. zOS is now able to compile libuv successfully.
The remaining changes are to do with failing tests and async io. These changes might be a little bit more invasive so I would prefer to create separate PRs for them. Let me know if this is a good idea.

**Comment by saghul** [comment]
Sounds good. I'll give this a review as soon as I can.

File: src/unix/udp.c [hunk 16]
```diff
@@ -550,6 +557,10 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
                  optname,
                  &mreq,
                  sizeof(mreq))) {
+#if defined(__MVS__)
+  if(errno == ENXIO)
```
Comment (cjihrig) [inline src/unix/udp.c]: Space after `if`.

File: src/unix/udp.c [hunk 17]
```diff
@@ -507,6 +510,10 @@ static int uv__udp_set_membership4(uv_udp_t* handle,
...
```
Comment (cjihrig) [inline src/unix/udp.c]: Here too.

File: src/unix/pipe.c [hunk 18]
```diff
@@ -97,6 +97,11 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
   if (uv__stream_fd(handle) == -1)
     return -EINVAL;
 
+#if defined(__MVS__)
+  /*on zOS, backlog=0 has undefined behaviour */
```
Comment (cjihrig) [inline src/unix/pipe.c]: Space after `/*`. Please capitalize "on".

File: include/uv-unix.h [hunk 19]
```diff
@@ -137,8 +137,17 @@ typedef pthread_rwlock_t uv_rwlock_t;
 typedef UV_PLATFORM_SEM_T uv_sem_t;
 typedef pthread_cond_t uv_cond_t;
 typedef pthread_key_t uv_key_t;
-typedef pthread_barrier_t uv_barrier_t;
```
Comment (saghul) [inline include/uv-unix.h]: this change should be part of the pthread-barrier compatibility shim. Check include/pthread-barrier.h and src/unix/pthread-barrier.c

File: src/unix/tcp.c [hunk 20]
```diff
@@ -115,6 +115,10 @@ int uv__tcp_bind(uv_tcp_t* tcp,
                    IPV6_V6ONLY,
                    &on,
                    sizeof on) == -1) {
+#if defined(__MVS__)
+      if (errno == EOPNOTSUPP)
+        return UV_EINVAL;
```
Comment (saghul) [inline src/unix/tcp.c]: return -EINVAL

File: src/unix/stream.c [hunk 21]
```diff
@@ -559,7 +559,11 @@ void uv__server_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
     if (stream->type == UV_TCP && (stream->flags & UV_TCP_SINGLE_ACCEPT)) {
       /* Give other processes a chance to accept connections. */
       struct timespec timeout = { 0, 1 };
+#if defined(__MVS__)
+      usleep(1);
```
Comment (saghul) [inline src/unix/stream.c]: can this ifdef be vaoided by using something which works everywhere?

File: src/unix/udp.c [hunk 22]
```diff
@@ -700,19 +711,24 @@ int uv_udp_set_ttl(uv_udp_t* handle, int ttl) {
   if (ttl < 1 || ttl > 255)
     return -EINVAL;
 
+#if defined(__MVS__)
+  if (!(handle->flags & UV_HANDLE_IPV6))
+    return -EINVAL;  /* zOS does not support setting ttl for IPv4 */
```
Comment (saghul) [inline src/unix/udp.c]: ENOTSUP instead?

**Comment by saghul** [comment]
Left another round of comments. The only important stuff is the pthread_barrier compatibility shim, once that's fixed I think we are ready to land.

File: include/uv-unix.h (hunk 19 above)
Comment (jBarz) [inline include/uv-unix.h]: I agree. I have removed this changeset.

Comment (saghul) [inline include/uv-unix.h]: Did you push? I still the changes here.

File: src/unix/stream.c (hunk 21 above)
Comment (jBarz) [inline src/unix/stream.c]: I found an implementation of nanosleep on z/OS. Will use that instead.

File: include/uv-unix.h (hunk 19 above)
Comment (jBarz) [inline include/uv-unix.h]: Just pushed.

File: src/unix/tcp.c [hunk 23]
```diff
@@ -115,6 +115,10 @@ int uv__tcp_bind(uv_tcp_t* tcp,
                    IPV6_V6ONLY,
                    &on,
                    sizeof on) == -1) {
+#if defined(__MVS__)
+      if (errno == EOPNOTSUPP)
+        return -EINVAL;
```
Comment (saghul) [inline src/unix/tcp.c]: i think it's better is we return ENOTSUP here

**Comment by saghul** [comment]
@jBarz one last minor nit, otherwise LGTM! Can we call libuv "Enterprise Ready" now? :smile:

File: src/unix/tcp.c (hunk 23 above)
Comment (jBarz) [inline src/unix/tcp.c]: Test "tcp_bind6_error_inval" expects EINVAL when binding a socket to an already bound loopback address. 
However, on z/OS the error instead of EINVAL is EOPNOTSUPP.
Do you think I should instead change the test case to expect EOPNOTSUPP on z/OS?

Comment (saghul) [inline src/unix/tcp.c]: oh, ok, leave it then.

**Comment by saghul** [comment]
LGTM. @libuv/collaborators if anyone plans to have a look let me know, otherwise I'll land this later today.

File: Makefile.am [hunk 24]
```diff
@@ -281,6 +285,23 @@ if SUNOS
...
+test_run_tests_CFLAGS += -D_UNIX03_THREADS \
+                         -D_UNIX03_SOURCE \
+                         -D_OPEN_SYS_IF_EXT=1 \
+                         -D_OPEN_SYS_SOCK_IPV6 \
+                         -D_OPEN_MSGQ_EXT \
+                         -D_XOPEN_SOURCE_EXTENDED \
+                         -D_ALL_SOURCE \
+                         -D_LARGE_TIME_API \
+                         -D_OPEN_SYS_FILE_EXT \
+                         -DUV_PLATFORM_SEM_T=int \
```
Comment (bnoordhuis) [inline Makefile.am]: UV_PLATFORM_SEM_T should be defined in include/uv-unix.h, not through a command line override.  Please remove it from the gyp file as well.

File: common.gypi [hunk 25]
```diff
@@ -11,7 +11,7 @@
     'configurations': {
       'Debug': {
         'defines': [ 'DEBUG', '_DEBUG' ],
-        'cflags': [ '-g', '-O0', '-fwrapv' ],
+        'cflags': [ '-g' ],
```
Comment (bnoordhuis) [inline common.gypi]: Why this seemingly unrelated change?

File: include/pthread-barrier.h [hunk 26]
```diff
@@ -18,7 +18,11 @@ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 #define _UV_PTHREAD_BARRIER_
 #include <errno.h>
 #include <pthread.h>
+#if defined( __MVS__)
+#define sem_t int
```
Comment (bnoordhuis) [inline include/pthread-barrier.h]: I don't think it's a good idea to pollute the global namespace like this.

File: src/unix/atomic-ops.h [hunk 27]
```diff
@@ -42,6 +42,11 @@ UV_UNUSED(static int cmpxchgi(int* ptr, int oldval, int newval)) {
   const int out = (*(volatile int*) ptr);
   __compare_and_swap(ptr, &oldval, newval);
   return out;
+#elif defined(__MVS__)
+  int out;
+  out = *ptr;
```
Comment (bnoordhuis) [inline src/unix/atomic-ops.h]: This is not atomic.

File: src/unix/pipe.c [hunk 28]
```diff
@@ -97,6 +97,11 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
   if (uv__stream_fd(handle) == -1)
     return -EINVAL;
 
+#if defined(__MVS__)
+  /* On zOS, backlog=0 has undefined behaviour */
+  backlog = backlog == 0 ? 1 : backlog;
```
Comment (bnoordhuis) [inline src/unix/pipe.c]: Can you write this as:
```c
if (backlog == 0)
  backlog = 1;
```
What about backlog < 0?

File: src/unix/thread.c [hunk 29]
```diff
@@ -302,6 +307,87 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+  buf.sem_op = value;
+  buf.sem_flg = 0;
+
+  semid = semget(IPC_PRIVATE, 1, S_IRUSR | S_IWUSR);
+  if (semid == -1)
+    return -errno;
+
+  op_status = semop(semid, &buf, 1); 
+  if (op_status == -1)
+    return -errno;
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Doesn't this leak the semaphore that is created a few lines up?

File: src/unix/thread.c [hunk 30]
```diff
@@ -302,6 +307,87 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+  return 0;
+}
+
+void uv_sem_destroy(uv_sem_t* sem) {
+  int op_status; 
+  op_status = semctl(*sem, 0, IPC_RMID);
+
+  if (op_status == -1)
+    abort();
+}
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Could be shortened too just:
```c
if (-1 == semctl(*sem, 0, IPC_RMID))
  abort();
```

File: src/unix/thread.c [hunk 31]
```diff
@@ -302,6 +307,87 @@ int uv_sem_trywait(uv_sem_t* sem) {
...
+  struct sembuf buf;
+  int op_status;
+
+  buf.sem_num = 0;
+  buf.sem_op = 1;
+  buf.sem_flg = 0;
+
+  op_status = semop(*sem, &buf, 1);
+  if (op_status == -1)
+    abort();
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Ditto.

File: src/unix/tty.c [hunk 32]
```diff
@@ -30,6 +30,10 @@
 #include <errno.h>
 #include <sys/ioctl.h>
 
+#if defined(__MVS__)
```
Comment (bnoordhuis) [inline src/unix/tty.c]: Perhaps this should be `#if defined(__MVS__) && !defined(IMAXBEL)` in case it gets defined in the future?

File: src/unix/udp.c [hunk 33]
```diff
@@ -700,19 +711,24 @@ int uv_udp_set_ttl(uv_udp_t* handle, int ttl) {
...
  */
-#if defined(__sun) || defined(_AIX) || defined(__OpenBSD__)
+#if defined(__sun) || defined(_AIX) || defined(__OpenBSD__) || defined(__MVS__)
   return uv__setsockopt(handle,
                         IP_TTL,
                         IPV6_UNICAST_HOPS,
                         &ttl,
                         sizeof(ttl));
-#endif /* defined(__sun) || defined(_AIX) || defined (__OpenBSD__) */
+#endif /* defined(__sun) || defined(_AIX) || defined (__OpenBSD__) || defined(__MVS__) */
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Long line.

File: test/runner-unix.c [hunk 34]
```diff
@@ -372,7 +376,11 @@ void process_cleanup(process_info_t *p) {
 
 /* Move the console cursor one line up and back to the first column. */
 void rewind_cursor(void) {
+#if defined(__MVS__)
+  fprintf(stderr, "\047[2K\r");
+#else
   fprintf(stderr, "\033[2K\r");
```
Comment (bnoordhuis) [inline test/runner-unix.c]: I'm surprised that doesn't work on zos.  What is the backstory?

File: test/test-udp-options.c [hunk 35]
```diff
@@ -52,7 +52,14 @@ static int udp_options_test(const struct sockaddr* addr) {
   /* values 1-255 should work */
   for (i = 1; i <= 255; i++) {
     r = uv_udp_set_ttl(&h, i);
+#if defined(__MVS__)
+    if(addr->sa_family == AF_INET6)
```
Comment (bnoordhuis) [inline test/test-udp-options.c]: Style: space before (.

**Comment by bnoordhuis** [comment]
@jBarz Can you explain why the UTMP_FILE define is necessary?

File: common.gypi (hunk 25 above)
Comment (jBarz) [inline common.gypi]: The options "-O0 -fwrapv" are not valid for the zOS compiler.
So I took those out and created the conditional below in the file.
+          ['OS != "zos"', {
+            'cflags': [ '-O0', '-fwrapv' ]
+          }],

Comment (bnoordhuis) [inline common.gypi]: Is it -O0 or -fwrapv that is not recognized?  Or is it both?

Comment (jBarz) [inline common.gypi]: Both of them are not recognized.

File: include/pthread-barrier.h (hunk 26 above)
Comment (jBarz) [inline include/pthread-barrier.h]: I have removed this. Instead I will define the following for z/OS.
#define UV_BARRIER_STRUCT_PADDING 0
For z/OS there is no need for binary compatibility with previous release because it was not enabled in the previous release :smile:

File: Makefile.am (hunk 24 above)
Comment (jBarz) [inline Makefile.am]: I will define in include/uv-os390.h

File: src/unix/atomic-ops.h (hunk 27 above)
Comment (jBarz) [inline src/unix/atomic-ops.h]: Will do
`const int out = (*(volatile int*) ptr);`

File: src/unix/pipe.c (hunk 28 above)
Comment (jBarz) [inline src/unix/pipe.c]: backlog < 0 is also undefined. I will write the following
```
if (backlog <= 0) 
  backlog = 1;
```

File: src/unix/thread.c (hunk 29 above)
Comment (jBarz) [inline src/unix/thread.c]: Yes it does. I will free up the resource before returning.

File: test/runner-unix.c (hunk 34 above)
Comment (jBarz) [inline test/runner-unix.c]: The character encoding on z/OS is "ebcdic" which is quite different from ascii.
In this case, we want to rewind using the "ESC" character which is (in octal) 33 in ascii and 47 in ebcdic.
Here is a cheatsheet for ascii ebcdic translation
[link](http://www.simotime.com/asc2ebc1.htm)

**Comment by jBarz** [comment]
I have removed the UTMP_FILE define as it is unnecessary.

File: src/unix/atomic-ops.h [hunk 36]
```diff
@@ -42,6 +42,10 @@ UV_UNUSED(static int cmpxchgi(int* ptr, int oldval, int newval)) {
   const int out = (*(volatile int*) ptr);
   __compare_and_swap(ptr, &oldval, newval);
   return out;
+#elif defined(__MVS__)
+  const int out = (*(volatile int*) ptr);
+  cs((cs_t*)&oldval, (cs_t*)ptr, *(cs_t*)(&newval));
```
Comment (bnoordhuis) [inline src/unix/atomic-ops.h]: It still isn't fully atomic because there is a time window between the assignment to `out` and the compare-and-swap.  I guess it's acceptable for now because AIX  has the same bug but it should be fixed someday.

Comment (jBarz) [inline src/unix/atomic-ops.h]: I will look for a fix soon.

**Comment by saghul** [comment]
Ping, what's the state of this?

**Comment by jBarz** [comment]
This should be good to integrate.

**Comment by saghul** [comment]
Great, let's do this.

**Comment by saghul** [comment]
Landed in 1cff5b7 after squashing and fixing a couple of whitespace issues. I guess we can call libuv Enterprise Ready! :smile: :tada: :tada: :tada:

**Comment by jBarz** [comment]
Thanks! Enterprise world is going to love libuv! :-)

**Comment by mhdawson** [comment]
Great to see this land :)
